### DISCUSSION TITLE
Writing rays to a file

### DISCUSSION THREAD

**Discussion by keithrse** [mitsuba-renderer/mitsuba3#733] [General]
As I understand how mitsuba's ray propagation works, Rays are instantiated by emitters, then traced using a specified technique. Each ray however, only exists for a single intersection at which point its either absorbed (or presumably can leave the scene), or reflected. In the former case the ray is dead and its active status is set to false (presumably) and in the latter a new ray is instantiated where its parent reflected with an updated position/direction. 

Question 1 : Am I on the right track regarding the mechanics?

The reason Im asking, is because what I need from mitsuba is the final position (final state) of each ray when it was absorbed, reached its max depth or in general was killed. I want the final states before they've undergone any MC integration.

Question 2 : Is it possible to extract this information, and if so where are rays killed?

**Comment by njroussel** [comment]
Hi @keithrse 

1) Yes, this is gist of how an `AdjointIntegrator` like the `ptracer` works. There are a few more cases where a ray can become inactive, like during the Russian Roulette mechanism (rays are randomly made inactive).

2) Each integrator defines how and when it makes rays inactive. For the `ptracer` specifically it is these lines of code:https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/integrators/ptracer.cpp#L265-L267. In the same loop body, you'll see a few other places where the `active` variable is modified as I mentioned earlier. Saving/extracting these requires some work, it's not something the interface was planned for. 

In general, if you're using a JIT variant, trying to save an intermediate variable in the rendering loop is a costly operation as it will break the megakernel and split it every time you try to save the variable.
